### DISCUSSION TITLE
Difference with SQL query and search form.

### DISCUSSION THREAD

**Discussion by meteos77** [textbrowser/biblioteq#173] [General]
a small difference between the 2 results

Search "Mémoires" in location :  2 books in my database
select * from book where location like "Mémoires"  with the sql queryl : **2** books OK
with the form Ctrl+F  : **1** single result line and  empty line

other examples T-Théatre
select * from book where location like "T-Théatre" => Result **6** books => OK
with the form Ctrl+F : **5** books + empty line

my database test
[biblioteQ-2022-01-09-version4.sqlite.zip](https://github.com/textbrowser/biblioteq/files/7841016/biblioteQ-2022-01-09-version4.sqlite.zip)

**Comment by textbrowser** [comment]
The empty line is most likely attached irrespective of the query.

**Comment by textbrowser** [comment]
OK, corrected optimization.

**Reply by meteos77** [reply]
for the optimization SUPER thank you again
I start to test in real condition

**Comment by meteos77** [comment]
problem when searching by accession number ; id ; title ...

**Reply by textbrowser** [reply]
Works here. Need more details.

**Comment by meteos77** [comment]
I test with my small base with the covers 2021-11-27_liste-47_achat_nouveautés.sqlite  in pj in subject new option : Print Icons View 

by accession_number : 7056 -> empty result
by title : BETTY --> empty result

**Reply by textbrowser** [reply]
Still not clear. There is no Print Icons View By...

**Comment by meteos77** [comment]
sorry, I start again from the beginning
test base
[2021-11-27_liste-47_achat_nouveautés.sqlite.zip](https://github.com/textbrowser/biblioteq/files/7847515/2021-11-27_liste-47_achat_nouveautes.sqlite.zip)

**Comment by meteos77** [comment]
I am looking for the book that has the access number 7093

**Comment by meteos77** [comment]
I am looking for the book with the title : BETTY

with the version of 15/12/2021 and the same base it is OK

**Comment by textbrowser** [comment]
The optimization is the problem?

**Reply by meteos77** [reply]
I filled in the category field with "AB" on 4 books, if I do the search
it finds me 3 then 4

accession number for example : all the documents that start with 7
should find 47, it finds 46

I have the impression that there is a cache that does not update automatically

**Reply by meteos77** [reply]
precision if it can put you on a track: it is the last of the list: the 7102

**Comment by meteos77** [comment]
the last one is always missing and so when there is only one then there is nothing!

**Reply by meteos77** [reply]
confirmation with category test there is always the total -1

**Comment by textbrowser** [comment]
I think it's now correct. I removed some of the previous operations as they were confusing and incorrect. They also did not provide performance.

**Reply by meteos77** [reply]
nice work, it works as before with high speed :-)

**Comment by meteos77** [comment]
I noticed a new slight malfunction the "case insensitive" in the search seems inoperative ?

that I type betty or BETTY, it finds me the book, case checked or not

I don't mind at all, it's rather the opposite, is there any characters to make the search like a light regex  * .   ?

I answer myself for my doc
_ for 1 character
% for all characters

**Comment by textbrowser** [comment]
You may be correct with the sensitivity. If I remember correctly, there is a special string function which assumes a default of false so if you wish to perform a case-sensitive search, the request is ignored.

**Reply by meteos77** [reply]
that's fine, it was to point out

**Comment by textbrowser** [comment]
Batch one of corrections completed! :)

**Comment by textbrowser** [comment]
Very peculiar SQLite behavior.

> SELECT title FROM book WHERE title LIKE '%betty%';

One row is returned.

**Comment by textbrowser** [comment]
This is an SQLite-ism. It functions correctly with PQ.

https://www.sqlite.org/pragma.html

PRAGMA case_sensitive_like = boolean;

The default behavior of the LIKE operator is to ignore case for ASCII characters. Hence, by default 'a' LIKE 'A' is true. The case_sensitive_like pragma installs a new application-defined LIKE function that is either case sensitive or insensitive depending on the value of the case_sensitive_like pragma. When case_sensitive_like is disabled, the default LIKE behavior is expressed. When case_sensitive_like is enabled, case becomes significant. So, for example, 'a' LIKE 'A' is false but 'a' LIKE 'a' is still true.

This pragma uses sqlite3_create_function() to overload the LIKE and GLOB functions, which may override previous implementations of LIKE and GLOB registered by the application. This pragma only changes the behavior of the SQL LIKE operator. It does not change the behavior of the sqlite3_strlike() C-language interface, which is always case insensitive.

**Comment by textbrowser** [comment]
One query adjusted! A trillion left!

**Comment by textbrowser** [comment]
Both general searches for SQLite now have the special case-sensitive-like option enabled.

**Reply by meteos77** [reply]
it works well;  I would have done the opposite but I don't know the history of this option.

put as text : "case-**sensitive** search" instead of "case-**insensitive** search". 
and uncheck by default because otherwise by default it may return nothing :-) and the one who wants to refine his search he specifies case-sensitive and modify the searched text.

**Comment by textbrowser** [comment]
There are times when I search for specific yet repeated text that I wish the case sensitive search would be enabled instead of the default case insensitive search. Hence, the option is case sensitive. My assumption is that for a library, precise searches have more importance than general searches.

**Reply by meteos77** [reply]
for me the search we can do is on the title (impossible to know the case in advance; title; TITLE; Title (ex: 3774; 3121) in my database > 7000 documents
or the authors : name, NAME ; Name
or the subject (category) : there it is even more the lottery.

maybe provide a default preference like "accession_number Sort / Numeric/Text)

**Comment by textbrowser** [comment]
Someday in the future the machines will be able to ascertain your search patterns. And after that they'll be able to enable the sensitivity by grasping the data beforehand and either randomizing the sensitivity, a mixture of your patterns and data quantity, or based on data alone.

**Reply by meteos77** [reply]
why does biblioteQ not predict 1st search case sensitive and if result zero then it switches to case insensitive and does a new search

**Reply by textbrowser** [reply]
If you're on a WiFi network over the land of PQ, the connection quality is not generous. The state of the connection is also another thing. For  another project, I created a connection watcher for PQ. If the connection is lost or dead, BQ will not be made aware. I don't know if there are now things which can notify you of a disconnect event.

**Comment by textbrowser** [comment]
Because that is an expensive thing especially for a large database.

**Reply by meteos77** [reply]
but the user who gets zero results, will he do a new search ?

**Comment by textbrowser** [comment]
Are you asking if zero results persuades someone to stop searching? Yes.

**Reply by meteos77** [reply]
or not :-)

**Comment by textbrowser** [comment]
Critical thinkers are required.

**Reply by meteos77** [reply]
I can't stay I have to go away for a while
but a librarian talks with a member about a book but doesn't know the spelling of the title or the author's name, so he's bound to know that there's something to find in the database
human memory / machine memory

we have the same problem with the old software, people don't know the wildcards

**Comment by textbrowser** [comment]
On the likeliness of LIKE, there are also language nuances. I may be incorrect, so additional verification is required. LIKE for PQ and SQ may not be correct for all supported human languages.

**Reply by meteos77** [reply]
good luck with the non-Latin languages :-)

**Comment by textbrowser** [comment]
It is locale. If the locale is incorrect, oh no. Well, I have been contacted by several Greek participants and they're fine with the product. You have offered the most detailed inspection.

**Reply by meteos77** [reply]
it's great to have correspondents in the 4 corners of the world, it allows real life tests.
my tests are simple, I go from menu to menu and I look at the options :-)
my requests are related to the daily life of the library. I don't find the way to do it in BiblioteQ, I open a discussion to have the functionality.

I tried to put the Arabic language with the problem raised ... I do not know how you manage to solve the problems when you change alphabet in any case bravo.

**Comment by textbrowser** [comment]
State now retained for the general search. Not the basic search, the general one.

**Reply by meteos77** [reply]
it works well. why the option does not concern the search line of the table?

**Reply by textbrowser** [reply]
The basic search? That requires a separate (key, value) pair. Never store one pair for multiple events. It violates many principles which I've ignored happily.

**Comment by textbrowser** [comment]
https://en.wikipedia.org/wiki/Microm%C3%A9gas

**Reply by meteos77** [reply]
you are interested in French classical literature :-)
in my database of > 7000 documents : accession_number : 1579 (Micromégas) so it is in the library.
I would read it if it is not written too small because generally in these collections the size is too small for my eyes :-(
for the search : Micromégas becomes MICROMEGAS so the search became delicate by respecting the case because the correct spelling is MICROMÉGAS or Micromégas. (the accents disappeared from most of the titles because of the lack of rigor during the typing and the lack of facility of the keyboards to propose the accented capital letters at the time).

do you think that with the new telescope "James Webb" we will finally see the others ?

**Comment by textbrowser** [comment]
I prefer Russian as it does not evade reality.

**Comment by textbrowser** [comment]
Not if the Dark Forest Theory is true.

**Reply by textbrowser** [reply]
https://en.wikipedia.org/wiki/Fermi_paradox#Communication_is_dangerous

**Reply by meteos77** [reply]
it's an earthy vision :-)

**Reply by meteos77** [reply]
not a human vision

**Reply by meteos77** [reply]
a previous topic: sharing culture in rural areas)

here is the interest of our library Departmental of Loans (BDP) which provides us books with the file unimarc 995
you talk to me about a book, I don't have it but I can order it and in 1 month maximum I get it if it is available

there are 3 volumes, do you advise the 3 in order or just the one of the black forest theory 

![LiuCixin](https://user-images.githubusercontent.com/56431039/149174944-9f829012-d1e7-4582-ba1a-01067a79e247.png)
?

**Comment by textbrowser** [comment]
Aha! You presented a new idea. Anonymous reservation statistics!

**Reply by meteos77** [reply]
YES we have to provide the BDP with annual statistics on borrowing: but I don't know yet what kind, I have to ask our correspondent.
I try to have the form of last year to specify the requests

I suppose he would like to know the statistics 

- on the type of book borrowed (Novel, detective novel,
- on the type of Documentary (Dewey classification (000 100 200 300 ...)  : select count (deweynumber) from book where location like "Classe 100 %"
- on the type of public adult teen youth ...
- on the borrowing time
- the number of people per day of operation

the list can be long
this will give me a little exercise SQL :-)

the kind of statistics that allow them to direct their purchases to the demands of the network libraries.

**Comment by textbrowser** [comment]
Yes, the complete volume.

**Reply by meteos77** [reply]
ordered :-)
